node/src/components/block_proposer.rs

```diff
@@ -380,33 +386,72 @@ impl BlockProposerReady {
+        let block_gas_limit = Gas::from(deploy_config.block_gas_limit);
 
         let mut transfers = Vec::new();
         let mut wasm_deploys = Vec::new();
+        let mut block_gas_running_total = Gas::zero();
+        let mut block_size_running_total = 0usize;
 
         for (hash, deploy_type) in self.sets.pending.iter() {
-            let under_max_transfers = transfers.len() < max_transfers;
-            let under_max_deploys = wasm_deploys.len() < max_deploys;
-            if !under_max_deploys && !under_max_transfers {
+            let at_max_transfers = transfers.len() == max_transfers;
+            let at_max_deploys = wasm_deploys.len() == max_deploys;
+
+            // Early exit if block limits are met.
+            // TODO: break early if gas total is met, but only once we have a constant for the cost
+            // of wasm-less transfers. if block_gas_running_total >= block_gas_limit
+            if block_size_running_total + DEPLOY_APPROX_MIN_SIZE >= max_block_size_bytes
+                || (at_max_deploys && at_max_transfers)
+            {
                 break;
             }
 
-            if self.is_deploy_valid(
-                deploy_type.header(),
+            // Skip this deploy, searching instead for another that may fit the bill.
+            if block_size_running_total + deploy_type.size() > max_block_size_bytes {
+                continue;
+            }
+
+            if !self.is_deploy_valid(
+                &deploy_type.header(),
                 block_timestamp,
                 &deploy_config,
                 &past_deploys,
-            ) && !past_deploys.contains(hash)
-                && !self.sets.finalized_deploys.contains_key(hash)
+            ) || past_deploys.contains(hash)
+                || self.sets.finalized_deploys.contains_key(hash)
+            {
+                continue;
+            }
+
+            let payment_amount_gas = match Gas::from_motes(deploy_type.payment_amount(), CONV_RATE)
             {
-                // all deploys in pending that aren't in finalized blocks or
-                // proposed blocks from the set `past_blocks`
-                if under_max_transfers && matches!(deploy_type, DeployType::Transfer(_)) {
-                    transfers.push(*hash)
-                } else if under_max_deploys && matches!(deploy_type, DeployType::Wasm(_)) {
-                    wasm_deploys.push(*hash)
+                Some(value) => value,
+                None => {
+                    tracing::error!("payment_amount couldn't be converted from motes to gas");
+                    continue;
                 }
+            };
+
+            let gas_running_total = if let Some(gas_running_total) =
+                block_gas_running_total.checked_add(payment_amount_gas)
+            {
+                gas_running_total
+            } else {
+                tracing::warn!("block gas would overflow");
+                break;
```

Comment:
If I submit a deploy with payment amount `U512::MAX`, won't that always cause us to `break` here, even if the block is still empty?
Would it make sense to just `continue` instead?

Reply:
I think that makes more sense, rather than allowing the proposer's behaviour to be affected by sending a max value.